examples/amm/src/contract.rs

```diff
@@ -0,0 +1,206 @@
+#![cfg_attr(target_arch = "wasm32", no_main)]
+
+mod state;
+
+use self::state::Amm;
+use amm::AmmError;
+use async_trait::async_trait;
+use linera_sdk::{
+    base::{SessionId, WithContractAbi},
+    ApplicationCallResult, CalleeContext, Contract, ExecutionResult, MessageContext,
+    OperationContext, SessionCallResult, ViewStateStorage,
+};
+
+linera_sdk::contract!(Amm);
+
+impl WithContractAbi for Amm {
+    type Abi = amm::AmmAbi;
+}
+
+#[async_trait]
+impl Contract for Amm {
+    type Error = AmmError;
+    type Storage = ViewStateStorage<Self>;
+
+    async fn initialize(
+        &mut self,
+        _context: &OperationContext,
+        _argument: (),
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        self.balance0.set(0);
+        self.balance1.set(0);
+        self.total_shares.set(0);
+        Ok(ExecutionResult::default())
+    }
+
+    async fn execute_operation(
+        &mut self,
+        _context: &OperationContext,
+        operation: Self::Operation,
+    ) -> Result<ExecutionResult<Self::Message>, AmmError> {
+        match operation {
+            Self::Operation::Swap {
+                input_token_idx,
+                output_token_idx,
+                input_amount,
+            } => {
+                if input_token_idx == output_token_idx {
+                    return Err(AmmError::EqualTokensError);
+                }
+
+                let input_pool_balance = *self.get_pool_balance(input_token_idx)?;
+                let output_pool_balance = *self.get_pool_balance(output_token_idx)?;
+
+                let output_amount = amm::calculate_output_amount(
+                    input_amount,
+                    input_pool_balance,
+                    output_pool_balance,
+                )?;
+
+                self.decrease_pool_balance(input_token_idx, input_amount)?;
+                self.increase_pool_balance(output_token_idx, output_amount)?;
+
+                Ok(ExecutionResult::default())
+            }
+            Self::Operation::AddLiquidity {
+                token0_amount,
+                token1_amount,
+            } => {
+                let balance0 = self.balance0.get_mut();
+                let balance1 = self.balance1.get_mut();
+
+                let balance0_float = *balance0 as f64;
+                let balance1_float = *balance1 as f64;
+
+                let token0_amount_float = token0_amount as f64;
+                let token1_amount_float = token1_amount as f64;
+
+                let total_shares = self.total_shares.get_mut();
+                let total_shares_float = *total_shares as f64;
+
+                if *total_shares > 0
+                    && balance0_float / balance1_float != token0_amount_float / token1_amount_float
+                {
+                    return Err(AmmError::BalanceRatioAlteredError);
```

Comment:
Ah, I only now saw that this was also @jvff's request.
But as he already wrote, we _have_ to allow a margin of error, otherwise it might simply be impossible to match the exact ratio.